camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlMultiOutputRecordWriter.java

```diff
@@ -68,7 +70,17 @@ public void write(EtlKey key, Object val) throws IOException, InterruptedExcepti
         //TODO: fix this logging message, should be logged once as a total count of old records skipped for each topic
         // for now, commenting this out
         //log.warn("Key's time: " + key + " is less than beginTime: " + beginTimeStamp);
-        ((Mapper.Context)context).getCounter("total", "skip-old").increment(1);
+        try {
+          //In Hadoop 2, TaskAttemptContext.getCounter() is available
+          //In Hadoop 2, TaskAttemptContextImpl cannot be cast to Mapper.Context
+          Method getCounterMethod = context.getClass().getMethod("getCounter", String.class, String.class);
+          ((Counter) getCounterMethod.invoke(context, "total", "skip-old")).increment(1);
+        } catch (NoSuchMethodException e) {
+          //In Hadoop 1, TaskAttemptContext.getCounter() is not available, has to cast context to Mapper.Context
+          ((Mapper.Context)context).getCounter("total", "skip-old").increment(1);
+        } catch (Exception e) {
```

Comment:
Does the instanceOfInterface mechanism work?

Why do we need to catch Exception here?  If we catch Exception here for hadoop2, we would need to do the same for hadoop1.

The old code doesn't capture exception, it's actually good in the sense this error wasn't lost in the map log.  If we capture the exception, we would need to record how many times we see the error and fail the job at the end.